flink-autoscaler/src/main/java/org/apache/flink/autoscaler/config/AutoScalerOptions.java

```diff
@@ -247,21 +222,41 @@ private static ConfigOptions.OptionBuilder autoScalerConfig(String key) {
                     .stringType()
                     .asList()
                     .defaultValues()
-                    .withFallbackKeys(oldOperatorConfigKey("vertex.exclude.ids"))
                     .withDescription(
                             "A (semicolon-separated) list of vertex ids in hexstring for which to disable scaling. Caution: For non-sink vertices this will still scale their downstream operators until https://issues.apache.org/jira/browse/FLINK-31215 is implemented.");
 
     public static final ConfigOption<Duration> SCALING_EVENT_INTERVAL =
             autoScalerConfig("scaling.event.interval")
                     .durationType()
                     .defaultValue(Duration.ofMinutes(30))
-                    .withFallbackKeys(oldOperatorConfigKey("scaling.event.interval"))
                     .withDescription("Time interval to resend the identical event");
 
     public static final ConfigOption<Duration> FLINK_CLIENT_TIMEOUT =
             autoScalerConfig("flink.rest-client.timeout")
                     .durationType()
                     .defaultValue(Duration.ofSeconds(10))
-                    .withFallbackKeys(oldOperatorConfigKey("flink.rest-client.timeout"))
                     .withDescription("The timeout for waiting the flink rest client to return.");
+
+    /** Migrate config keys still prefixed with the old Kubernetes operator prefix. */
+    public static Configuration migrateOldConfigKeys(Configuration config) {
+        Preconditions.checkNotNull(config);
+        config = new Configuration(config);
+
+        Set<String> toBeMigrated = new HashSet<>();
+        for (String key : config.keySet()) {
+            if (key.startsWith(LEGACY_CONF_PREFIX)) {
+                toBeMigrated.add(key);
+            }
+        }
+        for (String key : toBeMigrated) {
+            String migratedKey = key.substring(LEGACY_CONF_PREFIX.length());
+            boolean keyDoesNotExist = config.getString(migratedKey, null) == null;
+            if (keyDoesNotExist) {
+                String migratedValue = Preconditions.checkNotNull(config.getString(key, null));
+                config.setString(migratedKey, migratedValue);
+            }
+            config.removeKey(key);
+        }
+        return config;
+    }
```

Comment:
```suggestion
    public static Configuration migrateOldConfigKeys(Configuration config) {
        Map<String, String> oldConfig = Preconditions.checkNotNull(config).toMap();
        Configuration newConfig = new Configuration();

        for (Map.Entry<String, String> entry : oldConfig.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            if (!key.startsWith(LEGACY_CONF_PREFIX)) {
                newConfig.setString(key, value);
                continue;
            }
            String migratedKey = key.substring(LEGACY_CONF_PREFIX.length());
            if (newConfig.getString(migratedKey, null) == null) {
                newConfig.setString(migratedKey, value);
            }
        }

        return newConfig;
    }
```

How about this? Only one for loop.

Reply:
The O(n) complexity of both approaches is the same. I still prefer the first commit in this PR.

Reply:
There are a lot of copy steps with this approach because `toMap()` converts every single value before it inserts it into the new map. `new Configuration(config)` directly passes on the internal map. 

With that in mind, I tried incorporating your code suggestions, i.e. do a single pass.